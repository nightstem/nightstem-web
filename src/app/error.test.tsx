/* eslint-disable no-console */
import { render, screen } from '@testing-library/react';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import userEvent from '@testing-library/user-event';

import ErrorPage, { type ErrorPageProps, metadata } from '@/app/error';

describe('ErrorPage', () => {
  const mockReset = vi.fn();
  const mockError = new Error('Test error message');

  const defaultProps: ErrorPageProps = {
    error: mockError,
    reset: mockReset,
  };

  const originalConsoleError = console.error;

  beforeEach(() => {
    console.error = vi.fn();
  });

  afterEach(() => {
    console.error = originalConsoleError;
  });

  it('does not have any accessibility violations', async () => {
    const { container } = render(<ErrorPage {...defaultProps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('renders correctly with snapshot', () => {
    const { container } = render(<ErrorPage {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('Behavior', () => {
    it('logs error to console on mount', () => {
      render(<ErrorPage {...defaultProps} />);

      expect(console.error).toHaveBeenCalledWith(mockError);
      expect(console.error).toHaveBeenCalledTimes(1);
    });

    it('logs new error when error prop changes', () => {
      const { rerender } = render(<ErrorPage {...defaultProps} />);

      const newError = new Error('New error message');
      rerender(<ErrorPage error={newError} reset={mockReset} />);

      expect(console.error).toHaveBeenCalledWith(mockError);
      expect(console.error).toHaveBeenCalledWith(newError);
      expect(console.error).toHaveBeenCalledTimes(2);
    });

    it('calls reset function when reset button is clicked', async () => {
      const user = userEvent.setup();
      render(<ErrorPage {...defaultProps} />);

      const resetButton = screen.getByRole('button', { name: 'Try again' });
      await user.click(resetButton);

      expect(mockReset).toHaveBeenCalledTimes(1);
    });

    it('handles error with digest property', () => {
      const errorWithDigest = Object.assign(new Error('Test error'), {
        digest: 'error-digest-123',
      });

      render(<ErrorPage error={errorWithDigest} reset={mockReset} />);

      expect(console.error).toHaveBeenCalledWith(errorWithDigest);
    });
  });
});

describe('metadata', () => {
  it('exports correct metadata object', () => {
    expect(metadata).toBeDefined();
    expect(metadata).toEqual({
      title: 'Unexpected error | Nightstem',
      robots: 'noindex',
    });
  });

  it('has correct title for SEO', () => {
    expect(metadata.title).toBe('Unexpected error | Nightstem');
  });

  it('has noindex robots directive', () => {
    expect(metadata.robots).toBe('noindex');
  });
});
