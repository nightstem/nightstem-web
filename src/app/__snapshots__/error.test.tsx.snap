// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorPage > Rendering > renders correctly with snapshot 1`] = `
<div
  class="h-full w-full grid place-items-center"
>
  <div
    aria-describedby="nf-desc"
    aria-labelledby="nf-title"
    class="grid gap-8 w-full max-w-prose px-6"
  >
    <header
      class="grid gap-2"
    >
      <div
        class="inline-flex items-center gap-2"
      >
        <svg
          aria-hidden="true"
          class="text-primary-500 h-[1.066rem] w-[1.066rem] animate-pulse motion-reduce:animate-none shrink-0"
          fill="none"
          viewBox="0 0 500 500"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M346.04 135.875L500 245.427L161.185 495.252C157.629 497.874 151.828 497.923 148.228 495.362L0.786259 390.448L346.04 135.875Z"
            fill="currentColor"
          />
          <path
            d="M148.322 4.74771C151.878 2.12562 157.68 2.07657 161.279 4.63814L333.785 127.386L179.023 241.499L9.54552e-05 114.113L148.322 4.74771Z"
            fill="currentColor"
          />
        </svg>
        <span
          class="heading-sm"
        >
          Nightstem
        </span>
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        class="grid gap-8"
      >
        <h1
          class="heading-lg"
          id="nf-title"
        >
          Something went wrong
        </h1>
        <p
          class="text-sm md:text-base text-foreground/60"
          id="nf-desc"
        >
          An unexpected error occurred. Don't worry, we're on it.
        </p>
      </div>
    </header>
    <div
      class="flex gap-3"
    >
      <button
        class="select-none whitespace-nowrap font-medium cursor-pointer disabled:cursor-not-allowed transition-colors duration-200 motion-reduce:transition-none focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-dark text-base px-4 py-2 rounded-md text-dark bg-primary-500 hover:bg-primary-400 active:bg-primary-400 focus-visible:ring-primary-500 disabled:opacity-60 disabled:hover:bg-primary-500 disabled:active:bg-primary-500"
        type="button"
      >
        Try again
      </button>
    </div>
  </div>
</div>
`;

exports[`ErrorPage > Rendering > renders with error containing digest 1`] = `
<div
  class="h-full w-full grid place-items-center"
>
  <div
    aria-describedby="nf-desc"
    aria-labelledby="nf-title"
    class="grid gap-8 w-full max-w-prose px-6"
  >
    <header
      class="grid gap-2"
    >
      <div
        class="inline-flex items-center gap-2"
      >
        <svg
          aria-hidden="true"
          class="text-primary-500 h-[1.066rem] w-[1.066rem] animate-pulse motion-reduce:animate-none shrink-0"
          fill="none"
          viewBox="0 0 500 500"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M346.04 135.875L500 245.427L161.185 495.252C157.629 497.874 151.828 497.923 148.228 495.362L0.786259 390.448L346.04 135.875Z"
            fill="currentColor"
          />
          <path
            d="M148.322 4.74771C151.878 2.12562 157.68 2.07657 161.279 4.63814L333.785 127.386L179.023 241.499L9.54552e-05 114.113L148.322 4.74771Z"
            fill="currentColor"
          />
        </svg>
        <span
          class="heading-sm"
        >
          Nightstem
        </span>
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        class="grid gap-8"
      >
        <h1
          class="heading-lg"
          id="nf-title"
        >
          Something went wrong
        </h1>
        <p
          class="text-sm md:text-base text-foreground/60"
          id="nf-desc"
        >
          An unexpected error occurred. Don't worry, we're on it.
        </p>
      </div>
    </header>
    <div
      class="flex gap-3"
    >
      <button
        class="select-none whitespace-nowrap font-medium cursor-pointer disabled:cursor-not-allowed transition-colors duration-200 motion-reduce:transition-none focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-dark text-base px-4 py-2 rounded-md text-dark bg-primary-500 hover:bg-primary-400 active:bg-primary-400 focus-visible:ring-primary-500 disabled:opacity-60 disabled:hover:bg-primary-500 disabled:active:bg-primary-500"
        type="button"
      >
        Try again
      </button>
    </div>
  </div>
</div>
`;

exports[`ErrorPage > renders correctly with snapshot 1`] = `
<div
  class="h-full w-full grid place-items-center"
>
  <div
    aria-describedby="nf-desc"
    aria-labelledby="nf-title"
    class="grid gap-8 w-full max-w-prose px-6"
  >
    <header
      class="grid gap-2"
    >
      <div
        class="inline-flex items-center gap-2"
      >
        <svg
          aria-hidden="true"
          class="text-primary-500 h-[1.066rem] w-[1.066rem] animate-pulse motion-reduce:animate-none shrink-0"
          fill="none"
          viewBox="0 0 500 500"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M346.04 135.875L500 245.427L161.185 495.252C157.629 497.874 151.828 497.923 148.228 495.362L0.786259 390.448L346.04 135.875Z"
            fill="currentColor"
          />
          <path
            d="M148.322 4.74771C151.878 2.12562 157.68 2.07657 161.279 4.63814L333.785 127.386L179.023 241.499L9.54552e-05 114.113L148.322 4.74771Z"
            fill="currentColor"
          />
        </svg>
        <span
          class="heading-sm"
        >
          Nightstem
        </span>
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        class="grid gap-8"
      >
        <h1
          class="heading-lg"
          id="nf-title"
        >
          Something went wrong
        </h1>
        <p
          class="text-sm md:text-base text-foreground/60"
          id="nf-desc"
        >
          An unexpected error occurred. Don't worry, we're on it.
        </p>
      </div>
    </header>
    <div
      class="flex gap-3"
    >
      <button
        class="select-none whitespace-nowrap font-medium cursor-pointer disabled:cursor-not-allowed transition-colors duration-200 motion-reduce:transition-none focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 ring-offset-dark text-base px-4 py-2 rounded-md text-dark bg-primary-500 hover:bg-primary-400 active:bg-primary-400 focus-visible:ring-primary-500 disabled:opacity-60 disabled:hover:bg-primary-500 disabled:active:bg-primary-500"
        type="button"
      >
        Try again
      </button>
    </div>
  </div>
</div>
`;
